### PR TITLE
[LOOP-4349] Temporary mute alerts

### PR DESCRIPTION
**Pull request by nhamming**
https://tidepool.atlassian.net/browse/LOOP-4349

Related LoopKit PR: https://github.com/tidepool-org/LoopKit/pull/473

### CODE REVIEW

**Comment by nhamming** [comment]
@ps2 Again, consider this as an initial review of the concept. Design is still needed, which may trigger refinements. 

The concept is that there is an alert muter that determines when alerts should be muted. Muting of an alert happens by adding a muted flag to its metadata. Then when its sound is access to be played, those with this mute flag are set to .vibrate. This allows the sound that should be played to be retained incase the alert is unmuted before the mute duration expires.

File: Loop/Managers/AlertMuter.swift
Comment (ps2) [inline Loop/Managers/AlertMuter.swift]: Suggest `shouldMuteAlert(at alertDate: Date)`, as the caller is doing `timeIntervalSinceNow` on the actual alert date, and you're unwinding that here.  Simpler (and more intuitive imo), to just use the alertDate.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [inline Loop/Managers/Alerts/AlertManager.swift]: This seems to me to be using the wrong level of abstraction.  The alerts aren't retracted and then reissued. Any future-scheduled *issued* alerts remain issued, but are rescheduled with silence, like is being done above with the loop-not-looping alerts. This only impacts the system notifications, as they're the only ones that are "scheduled". Local modal presentation can check the AlertMuter state on presentation.

**Review by ps2 — COMMENTED** [review]
A couple of recommendations. I don't think the responsibility of muting should fall on the AlertManager, and alerts themselves do not need to change.

**Comment by nhamming** [comment]
@ps2 I believe this is at a point for another conceptual review. I'm in the act of updating unit tests, but I suspect there may need to be additional changes.

File: Loop/Managers/AlertMuter.swift
Comment (ps2) [inline Loop/Managers/AlertMuter.swift]: For some reason this wording is confusing to me. I think I understand it after reviewing the function. Maybe `func shouldMuteAlert(scheduledAt timeFromNow: TimeInterval = 0)` ?

File: Loop/Managers/AlertMuter.swift
Comment (ps2) [inline Loop/Managers/AlertMuter.swift]: "Off" in what sense? It should be fired immediately after Loop completed.

File: Loop/Managers/AlertMuter.swift
Comment (ps2) [inline Loop/Managers/AlertMuter.swift]: Rather than having a stateful enabled flag, that can be out of date if check() isn't called in a timely manner, it might be better to just have a start time that is set or not set, and have enabled return true if startDate is set, and now is < startDate + duration.

Comment (ps2) [reply Loop/Managers/AlertMuter.swift]: Then does `check()` even need to exist?

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [inline Loop/Managers/Alerts/AlertManager.swift:91]: Rescheduling only needs to happen in response to user action of enabling, or prematurely disabling, mute alert mode.  A natural expiration already does the right thing.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [inline Loop/Managers/Alerts/AlertManager.swift]: I believe `scheduleLoopNotRunningNotifications` assumes it is running immediately after the loop completes. If this assumption is violated, and it is called at other times, you may need to pass in lastLoopCompleted, so it can calculate the correct alerting times.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [inline Loop/Managers/Alerts/AlertManager.swift]: I think this just sets the sound to use default, not to actually mute the sound.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [inline Loop/Managers/Alerts/AlertManager.swift]: It would seem clearer to me to have the alert issuers know about the current mute state and do the right thing when issuing, rather than mutating alerts to mark them as muted or not.  This is reflecting my assumption that the "muting" we're implementing here is a property of the alert system, not a property of individual alerts.

File: Loop/Managers/Alerts/InAppModalAlertIssuer.swift
Comment (ps2) [inline Loop/Managers/Alerts/InAppModalAlertIssuer.swift:30]: I think this is good, but I think we need to adjust the `userNotificationCenter` `willPresent` delegate callback to allow notifications to go through while in foreground mode, so we get haptics, sounds, and watch delivery while in the foreground.

**Review by ps2 — COMMENTED** [review]
Thanks! I have a few bigger comments, so depending on where you go with them, we should probably have another review after this one.

File: Loop/Managers/AlertMuter.swift
Comment (nhamming) [reply Loop/Managers/AlertMuter.swift]: As long as `scheduledAt` is clear enough to imply a time interval from now, I'll use it.

File: Loop/Managers/AlertMuter.swift
Comment (nhamming) [reply Loop/Managers/AlertMuter.swift]: I'll re-word. Trying to clarify that the check related to a loop may change the state of the alert muter ~5 mins after it should have been changed (e.g., it was enabled right after a loop and would need to wait until 1 additional loop to be disabled).

File: Loop/Managers/Alerts/AlertManager.swift
Comment (nhamming) [reply Loop/Managers/Alerts/AlertManager.swift]: I'll test again, but from initial testing this mutes the sound, but not the vibration.

File: Loop/Managers/Alerts/InAppModalAlertIssuer.swift
Comment (nhamming) [reply Loop/Managers/Alerts/InAppModalAlertIssuer.swift:30]: yes, this was changed. See `LoopAppManager` line 447

File: Loop/Managers/Alerts/AlertManager.swift
Comment (nhamming) [reply Loop/Managers/Alerts/AlertManager.swift]: I started down this path but it requires a change to the AlertIssuer protocol that doesn't make sense for all AlertIssuers (cgm managers, pump managers, loop device manager, etc). All of those do not care of the system mute state, just that an alert needs to be issued. Only the `AlertManager` goes through the act of muting an alert. From our discussion of a system property, I thought this was in line.

Please suggest an alternative.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (nhamming) [reply Loop/Managers/Alerts/AlertManager.swift]: OK. I'll look and update.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (nhamming) [reply Loop/Managers/Alerts/AlertManager.swift:91]: Hm... no. Natural expiration of the duration leaves the state in the `enabled` state. There needs to be a check after the natural duration expires to trigger the state change to `disabled`. Or am I missing your point?

File: Loop/Managers/AlertMuter.swift
Comment (nhamming) [reply Loop/Managers/AlertMuter.swift]: Interesting... I'll see.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [reply Loop/Managers/Alerts/AlertManager.swift:91]: I mean a natural expiration doesn't require rescheduling.  See above for my suggestion on making `enabled` a computed var, which depends on startTime.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [reply Loop/Managers/Alerts/AlertManager.swift]: Yeah, I definitely agree about not changing the protocols used by sources of alerts (device managers, etc), including AlertIssuer.  I'm suggesting a change in the opposite direction, I think; I'm suggesting that `InAppModalAlertIssuer` and `UserNotificationAlertIssuer` are made aware of the mute state, and do the right thing, rather than having to tweak the alerts themselves, as is being done here. 

I'm guessing maybe you are hitting difficulties with AlertManager accessing `UserNotificationAlertIssuer` and `InAppModalAlertIssuer` via the `AlertIssuer` protocol. This has proved constraining to me in the past as well, and I think it might be worth revisiting how that relationship is modeled. Those two objects are really core parts of our alert system; they're not just generic "AlertIssuers" in the same sense of how devices use that api to issue alerts. I feel the relationship between AlertManager and those two classes is closer, and that representing that relationship as an unbounded array  (`handlers`) of `AlertIssuers` is maybe a premature abstraction.  When I tried to replace that with direct references, however, updating the tests proved difficult. You'll probably hit the same thing if you try it, but I'd welcome any steps in that direction.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (nhamming) [reply Loop/Managers/Alerts/AlertManager.swift:91]: Yes, I agree with no need to reschedule. However the UI needs to be updated to reflect the change. I'll look into using timers for this.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (nhamming) [reply Loop/Managers/Alerts/AlertManager.swift]: OK. I'll explore making the user notification alert issuer more explicit to the alert manager, since muting only needs to happen there.

**Comment by nhamming** [comment]
@ps2 First sorry for the force pushing. I know this makes continuing a review difficult. However, this branch has been around for a while and may require a complete review anyway.

I believe the issue for rescheduling loop not running notification is corrected. Also I added timers in UIs to force checking alert muter state for duration expiration.

File: Loop/Managers/AlertMuter.swift
Comment (ps2) [inline Loop/Managers/AlertMuter.swift]: I'm not sure I understand this comment. "disable the muting"? Does that mean it may play within a mute period?

File: Loop/Managers/AlertMuter.swift
Comment (ps2) [inline Loop/Managers/AlertMuter.swift]: Would `checkForMutePeriodEnding` be a valid name?  `check` is very vague.

File: Loop/Managers/AlertMuter.swift
Comment (ps2) [inline Loop/Managers/AlertMuter.swift]: maybe `isMuting`?

File: Loop/Managers/AlertMuter.swift
Comment (ps2) [inline Loop/Managers/AlertMuter.swift]: Was going to comment about "issued" date being a past tense verb, when it usually refers to a future alert, if set. But I see this naming issue goes deep, all the way to `PersistedAlert`.

File: Loop/Managers/AlertMuter.swift
Comment (ps2) [inline Loop/Managers/AlertMuter.swift]: So this method clears start time. What are the impacts if we just remove all these checks? `startTime` will remain set after the mute period ends, indefinitely. But `enabled` will return the correct value.

File: Loop/View Controllers/StatusTableViewController.swift
Comment (ps2) [inline Loop/View Controllers/StatusTableViewController.swift]: When the timer goes off, rather than having AlertMuter depend on UI components to make its state update, just having the UI component refresh (which should check `alertMuter.enabled`), seems like a cleaner separation of concerns.

**Review by ps2 — COMMENTED** [review]
Some more questions and I'm wondering if the whole `check` thing can go away.

File: Loop/Managers/AlertMuter.swift
Comment (nhamming) [reply Loop/Managers/AlertMuter.swift]: updated

File: Loop/Managers/AlertMuter.swift
Comment (nhamming) [reply Loop/Managers/AlertMuter.swift]: I ended up removing this.

File: Loop/Managers/AlertMuter.swift
Comment (nhamming) [reply Loop/Managers/AlertMuter.swift]: there needs to be a published update to trigger UI changes.

File: Loop/View Controllers/StatusTableViewController.swift
Comment (nhamming) [reply Loop/View Controllers/StatusTableViewController.swift]: I've updated this.

**Comment by nhamming** [comment]
@ps2 I made some updates before I looked at all your comments. Could you review again when you get a chance?

File: Loop/Managers/AlertMuter.swift
Comment (ps2) [inline Loop/Managers/AlertMuter.swift]: Debugging code.

File: Loop/Managers/AlertMuter.swift
Comment (ps2) [inline Loop/Managers/AlertMuter.swift:110]: Sorry, I meant that I was thinking it would be more appropriate for the *UI* components to own the timers. If there is nothing onscreen that needs to be updated, there is no need for timers.  The AlertMuter itself is able to return a correct value for isMuting without timers. I won't block on this, if you prefer this arrangement, as it should work fine.

**Review by ps2 — APPROVED** [review]
LGTM. Looks like there is some debug code left in. And another nonblocking comment about timer responsibility.

File: Loop/Managers/AlertMuter.swift
Comment (nhamming) [reply Loop/Managers/AlertMuter.swift:110]: I thought this would be a simpler solution. The UI may not know it needs a timer to track this since configuration is published.

**Comment by nhamming** [comment]
@ps2 looking for a formal review this time.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [inline Loop/Managers/Alerts/AlertManager.swift]: I know it's a weird spot to store this, but the "source of truth" last loop date is stored in `statusExtensionManager.context?.lastLoopCompleted`

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [inline Loop/Managers/Alerts/AlertManager.swift]: I thought the alert system was going to just reschedule any future scheduled/repeating alerts with the iOS notification system? This is retracting and re-issuing?

**Review by ps2 — COMMENTED** [review]
Main question I have is about "rescheduling" vs "retracting and reissuing".

File: Loop/Managers/Alerts/AlertManager.swift
Comment (nhamming) [reply Loop/Managers/Alerts/AlertManager.swift]: yes, but only with handlers. The alert store is not being touched in this case. Retraction removes the previous alert (muted or unmuted), say from a pending iOS notification, and issuing replaces it.

Curious of what you are considering/concerned with?

File: Loop/Managers/Alerts/AlertManager.swift
Comment (nhamming) [reply Loop/Managers/Alerts/AlertManager.swift]: yes, I know. We discussed putting the last loop date in the alert and this is from that discussion. Should it instead take it from the "source of truth"?

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [reply Loop/Managers/Alerts/AlertManager.swift]: Ok, if we're not actually "retracting" (like marking it in the database as retracted) and reissuing (with another entry in the database), then that is ok from a behavior standpoint, I think. But do we want to  be calling it "retract" and "issue", if this is just updating some signalization for the same alerts? 

And the modal components don't really need to be rescheduled, do they? Keeping the same timers should work, and when the timers fire, the mute/unmute state is evaluated.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [reply Loop/Managers/Alerts/AlertManager.swift]: Yeah, I do remember our discussion.  I think I was (incorrectly) thinking that these might have been scheduled with a "lastLoopTime" that was now forgotten.  The current tracking of last loop time should still be valid, in this case, and so storing it on the alerts is more complexity and redundant, I think.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (nhamming) [reply Loop/Managers/Alerts/AlertManager.swift]: unfortunately, for repeated alerts to keep the in-app and iOS notification in sync, both need to be retracted and issued.

I was reusing the existing language. But I'm happy to change it.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (nhamming) [reply Loop/Managers/Alerts/AlertManager.swift]: OK. I take this comment as `use the source of truth`. I'll make that update.

File: Loop/Managers/Alerts/AlertManager.swift
Comment (ps2) [reply Loop/Managers/Alerts/AlertManager.swift]: Ok, yeah, thanks for the reminder. It'd be nice if there was a way to update the sound for a pending iOS notification, but because we have to reschedule, we'll lose the original repeat offset, which means repeats will be out-of-sync with modal presentations. Sounds find to me.

**Review by ps2 — APPROVED** [review]
LGTM!

**Pull request merged**
